### PR TITLE
TypeScript

### PR DESCRIPTION
**Pull request by aminya**
# Ready to get merged - Tested and is working

### Description of the Change - Release Notes
- [X] Converting codebase source to typescript while being completely backward-compatible
- [x] Adding type definitions + documentation for API
- [x] Optimizing the code wherever possible

# Verification Process
- I verify the changes against other packages that use atom-select-list: for example, [atom-indent-detective](https://github.com/aminya/atom-indent-detective/tree/TypeScript) uses atom-select-list.
- The tests are untouched to show backward compatibility

### Possible Drawbacks
- Nothing (backward compatible - zero change in the API and JavaScript code) 

### Alternate Designs
- using @types/atom-select-list. But this doesn't allow us to use the power of typescript for optimizing and catching the bugs.

### CODE REVIEW

**Comment by aminya** [comment]
@maxbrunsfeld Could you review this?

**Comment by aminya** [comment]
Is there anyone to check this?

**Comment by aminya** [comment]
Is there anyone to check this? 

@nshikov @maxbrunsfeld @t9md @50Wliu

**Comment by aminya** [comment]
Could you check this PR? @lkashef @smashwilson  

How can I become a member of the `Atom` organization? I have at least 5 PRs that are open and no one reviews them. I can take responsibility for these and future potential inquiries.

File: package.json
Comment (winstliu) [inline package.json]: Can we just make this `tsc -p src/tsconfig.json` and not worry about hiding the output?

File: package.json
Comment (winstliu) [inline package.json:14]: Is `prepare` a special command or one that has some sort of conventional use (like build, test, etc)? What is its use here other than an alias to build?

File: src/select-list-properties.ts
Comment (winstliu) [inline src/select-list-properties.ts]: Missing documentation?

File: src/select-list-properties.ts
Comment (winstliu) [inline src/select-list-properties.ts]: Nit: I would wrap this line, and maybe some of the other longer ones as well.

File: src/select-list-view.ts
Comment (winstliu) [inline src/select-list-view.ts]: Can you elaborate on this comment (and the identical one in the constructor)?

File: src/select-list-view.ts
Comment (winstliu) [inline src/select-list-view.ts]: Can you change this back to the previous line-style comment?

File: src/select-list-view.ts
Comment (winstliu) [inline src/select-list-view.ts]: Ditto with the comment style.

File: src/select-list-view.ts
Comment (winstliu) [inline src/select-list-view.ts]:
```suggestion
  constructor (props: { element: EtchElement; selected: boolean; onclick: () => void }) {
```

Might be worth pulling this out into its own type for clarity.

File: src/select-list-view.ts
Comment (winstliu) [inline src/select-list-view.ts]: Oh yeah since we're using the same props type twice, definitely make it its own type.

File: src/tsconfig.json
Comment (winstliu) [inline src/tsconfig.json]: I don't believe this project uses decorators (please correct me if I'm wrong).

File: src/tsconfig.json
Comment (winstliu) [inline src/tsconfig.json]:
```suggestion
```
This is false by default.

File: src/tsconfig.json
Comment (winstliu) [inline src/tsconfig.json]: Can you go over why this is needed? A little out of touch with Atom's current ecosystem and debugging workflow.

File: src/tsconfig.json
Comment (winstliu) [inline src/tsconfig.json]: This shouldn't be necessary, right?

File: src/tsconfig.json
Comment (winstliu) [inline src/tsconfig.json]: This definitely shouldn't be needed considering what Etch is :P

File: src/tsconfig.json
Comment (winstliu) [inline src/tsconfig.json]: This is false by default.

File: src/tsconfig.json
Comment (winstliu) [inline src/tsconfig.json]: Can we instead move this into the root directory and specify `include` to only include the src folder, and `outDir` to be just `lib`?

File: src/tsconfig.json
Comment (winstliu) [inline src/tsconfig.json]: You've converted the one and only Javascript source file in this repo; why is this true?

File: src/tsconfig.json
Comment (winstliu) [inline src/tsconfig.json]: I've never seen this before, and it's also not showing up in the TSConfig reference. Can you explain its purpose?

File: src/select-list-view.ts
Comment (winstliu) [inline src/select-list-view.ts]: Can you elaborate on this comment?

File: src/select-list-view.ts
Comment (winstliu) [inline src/select-list-view.ts:1]: `atom` requires should come first

File: src/select-list-view.ts
Comment (winstliu) [inline src/select-list-view.ts]: Can this also be converted into an import for consistency? It would be either `import etch from 'etch'` or `import * as etch from 'etch'`, right?

File: package.json
Comment (winstliu) [inline package.json:9]: Hmm, might be worth adding a script to run `tsc watch`

File: src/tsconfig.json
Comment (winstliu) [inline src/tsconfig.json]: I'm uncomfortable with having all these safety checks set to false. Do we really fail all of these? How many are trivially fixable? For the ones that need more in-depth fixes, can comments be left as to why they're off?

**Review by winstliu — CHANGES_REQUESTED** [review]
Disclaimer: I'm no longer an Atom maintainer and my knowledge of the ecosystem is now 1.5 years old.
That said, still happy to give reviews that pop up on my notifications feed when I have the time.

This looks like a fairly straightforward port to Typescript. Have the current maintainers indicated that they're open to switching to Typescript? (I think there were some conversations about this...too fuzzy for me to recall for sure though)

Most of my comments are related to the configuration of Typescript itself, plus a few questions about the TODO comments you left.

File: package.json
Comment (aminya) [reply package.json]: I changed it in [this commit](https://github.com/atom/atom-select-list/pull/28/commits/a37549cde6322cd0d0745811a0e58a0b548f8358)

However, this means that we need to update the source code, which I wanted to avoid so it gets merged faster.

File: package.json
Comment (aminya) [reply package.json:14]: Prepare runs after local installation and before publishing. It is the recommended way to trigger the build.

https://docs.npmjs.com/cli/v6/using-npm/scripts#life-cycle-scripts

File: src/select-list-properties.ts
Comment (aminya) [reply src/select-list-properties.ts]: Yes. It is not documented in the readme, but the name seems self-explanatory.

File: src/select-list-properties.ts
Comment (aminya) [reply src/select-list-properties.ts]: Fixed in https://github.com/atom/atom-select-list/pull/28/commits/af6f304c7ae6072e8d253f0da918f6801a05b84d

File: src/select-list-view.ts
Comment (aminya) [reply src/select-list-view.ts]: `items` is identical in `props` and the constructor. So, for backward compatibility, I did not remove it. But it can be removed in a major version bump.

https://github.com/aminya/atom-select-list/blob/af6f304c7ae6072e8d253f0da918f6801a05b84d/src/select-list-view.ts#L39

File: src/select-list-view.ts
Comment (aminya) [reply src/select-list-view.ts]: Fixed in https://github.com/atom/atom-select-list/pull/28/commits/e5df058b6b014c6708eb6cb173ff25c8ef0cff4e

File: src/select-list-view.ts
Comment (aminya) [reply src/select-list-view.ts]: Fixed in https://github.com/atom/atom-select-list/pull/28/commits/e5df058b6b014c6708eb6cb173ff25c8ef0cff4e

File: src/select-list-view.ts
Comment (aminya) [reply src/select-list-view.ts]: Fixed in https://github.com/atom/atom-select-list/pull/28/commits/5abb4f0ef1e1a20ea56a4fa531da48de137718aa

File: src/select-list-view.ts
Comment (aminya) [reply src/select-list-view.ts]: Fixed in https://github.com/atom/atom-select-list/pull/28/commits/5abb4f0ef1e1a20ea56a4fa531da48de137718aa

File: src/tsconfig.json
Comment (aminya) [reply src/tsconfig.json]: I use this tsconfig template everywhere, so I don't have to be worried about changing it for every single project. But I can remove the unused parts.

Comment (aminya) [reply src/tsconfig.json]: Fixed all the tsconfig comments in this: https://github.com/atom/atom-select-list/pull/28/commits/4b5e63b73cce3f3ec560dfe1467a816dd20f93d9

File: src/tsconfig.json
Comment (aminya) [reply src/tsconfig.json]: Etch is written in JavaScript

File: src/select-list-view.ts
Comment (aminya) [reply src/select-list-view.ts]: Removed

File: src/select-list-view.ts
Comment (aminya) [reply src/select-list-view.ts:1]: Fixed in https://github.com/atom/atom-select-list/pull/28/commits/eaf175ddfce7bc0c96ab5c5324d3be3c991efa98

File: src/select-list-view.ts
Comment (aminya) [reply src/select-list-view.ts]: Fixed in https://github.com/atom/atom-select-list/pull/28/commits/eaf175ddfce7bc0c96ab5c5324d3be3c991efa98

File: package.json
Comment (aminya) [reply package.json:9]: Fixed in https://github.com/atom/atom-select-list/pull/28/commits/31f366586af3d3de564f6c353e13e5130870f3d8

File: src/tsconfig.json
Comment (aminya) [reply src/tsconfig.json]: There is no drawback in inlining the sources as far as I know.

File: src/tsconfig.json
Comment (aminya) [reply src/tsconfig.json]: I made tsconfig more strict

**Comment by aminya** [comment]
> Disclaimer: I'm no longer an Atom maintainer and my knowledge of the ecosystem is now 1.5 years old.
> That said, still happy to give reviews that pop up on my notifications feed when I have the time.
> 

Glad to hear from you!

> This looks like a fairly straightforward port to Typescript. Have the current maintainers indicated that they're open to switching to Typescript? (I think there were some conversations about this...too fuzzy for me to recall for sure though)
> 

There are only 2-3 people officially working part-time on Atom. They rarely get time to address the PRs. We are considering forking the whole organization in @atom-community, so I am not surprised anymore that I did not get any response here.

> Most of my comments are related to the configuration of Typescript itself, plus a few questions about the TODO comments you left.

I addressed most of them. I left some comments for the unresolved ones.

File: src/tsconfig.json
Comment (winstliu) [reply src/tsconfig.json]: Oof I mixed this up with checkJs. Sounds good 👍

File: src/select-list-view.ts
Comment (winstliu) [inline src/select-list-view.ts:438]: Doesn't need to be changed, but just wondering why you chose to use `type` here instead of an interface (though that may have been because I mentioned "type" in my comment 😁).

File: src/select-list-view.ts
Comment (winstliu) [inline src/select-list-view.ts]: Why the change?

File: tsconfig.json
Comment (winstliu) [inline tsconfig.json]: These two are implied with `strict`

File: src/select-list-view.ts
Comment (winstliu) [reply src/select-list-view.ts]: I looked through the code to re-familiarize myself with this. It technically is assigned in the constructor, via `computeItems`. Does Typescript throw an error if it's not explicitly set in the constructor itself? I forget how good it is at following the codeflow.

I'm not positive, but I think it is necessary because of the slices we do on it; this way `this.props.items` is the original copy and `this.items` is the filtered & ordered list that's shown.

File: src/select-list-view.ts
Comment (winstliu) [inline src/select-list-view.ts]:
```suggestion
```

File: src/select-list-view.ts
Comment (winstliu) [inline src/select-list-view.ts]: What's this TODO saying? And can an empty object really be passed in here?

File: src/select-list-properties.ts
Comment (winstliu) [inline src/select-list-properties.ts]: Seems like this should be optional as well

File: src/select-list-view.ts
Comment (winstliu) [reply src/select-list-view.ts]: And: is this meant to be public?

**Review by winstliu — COMMENTED** [review]
Mostly minor stuff, but still a few questions

File: src/select-list-view.ts
Comment (aminya) [reply src/select-list-view.ts]: This is a non-standard method, and it does not exist in the TypeScript types, if we want to keep it, we should declare the type for it.
https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoViewIfNeeded

File: tsconfig.json
Comment (aminya) [reply tsconfig.json]: Fixed in https://github.com/atom/atom-select-list/pull/28/commits/8818ab07a04e7b773560c543f634f5c9672fed9d

File: src/select-list-view.ts
Comment (aminya) [reply src/select-list-view.ts]: It is saying that we should choose better defaults instead of just passing `{}`

File: src/select-list-properties.ts
Comment (aminya) [reply src/select-list-properties.ts]: Fixed in https://github.com/atom/atom-select-list/pull/28/commits/e2bff3a3cac53b532efbb6e0d423ac815647e89c

**Comment by aminya** [comment]
> I looked through the code to re-familiarize myself with this. It technically is assigned in the constructor, via computeItems. Does Typescript throw an error if it's not explicitly set in the constructor itself? I forget how good it is at following the codeflow.

> I'm not positive, but I think it is necessary because of the slices we do on it; this way this.props.items is the original copy and this.items is the filtered & ordered list that's shown.

> And: is this meant to be public?

Yes, that seems like it. I will change it to private because it is not documented.

File: src/select-list-view.ts
Comment (winstliu) [reply src/select-list-view.ts]: Ahh...right. No, this still needs to be scrollIntoViewIfNeeded to avoid unexpected scrolling. Can this be ts-ignored instead of declaring the type? :D

File: src/select-list-view.ts
Comment (winstliu) [inline src/select-list-view.ts]:
```suggestion
  /** An array containing the filtered and ordered items to be shown in the select list. */
```

Would also remove the newline separating this one from all the other private variables.

File: src/select-list-view.ts
Comment (winstliu) [inline src/select-list-view.ts]: Did removing this not work?

File: src/select-list-view.ts
Comment (winstliu) [reply src/select-list-view.ts]: Per https://github.com/atom/etch/blob/18512fc64721d6f190e36415d81c09ec529d28ba/lib/patch.js#L46, it looks like `{}` _is_ the default value that's passed in if there's no new props. Which says two things:
1. The ` = {}` shouldn't even be needed b/c we should never be passed an undefined value, per the source above
2. Since it's intentional that we can be passed an empty object, I think it's fine that we just handle that normally.

So I don't think this TODO needs to stay.

File: src/select-list-view.ts
Comment (winstliu) [inline src/select-list-view.ts]: I think this type is EtchElement since it's coming from the children. _I think._

File: src/select-list-view.ts
Comment (winstliu) [inline src/select-list-view.ts:327]: Bit surprised Typescript doesn't complain about this now that it knows query is optional

**Review by winstliu — CHANGES_REQUESTED** [review]
Buncha small stuff. Only thing I'm blocking over is the last TODO comment.

(p.s. you'll do reviewers a favor if you elaborate more on what you mean by your TODOs in the future 🙂)

File: src/select-list-view.ts
Comment (aminya) [reply src/select-list-view.ts]: Removed in https://github.com/atom/atom-select-list/pull/28/commits/180f2240940f2911aa0703214e6803071f32f85b

File: src/select-list-view.ts
Comment (aminya) [reply src/select-list-view.ts]: We need to merge https://github.com/atom/etch/pull/90 if we want to fix all the types

File: src/select-list-view.ts
Comment (aminya) [reply src/select-list-view.ts:327]: It is because strictNullCheck is not true

**Comment by aminya** [comment]
> Per atom/etch@18512fc/lib/patch.js#L46, it looks like {} is the default value that's passed in if there's no new props. Which says two things:

> The = {} shouldn't even be needed b/c we should never be passed an undefined value, per the source above
Since it's intentional that we can be passed an empty object, I think it's fine that we just handle that normally.
So I don't think this TODO needs to stay.

Which one should I remove? `= {}` or the `TODO`?

**Comment by winstliu** [comment]
Both :)

File: src/select-list-view.ts
Comment (winstliu) [reply src/select-list-view.ts:327]: _Slightly_ concerning but it hasn't broken up until now, so let's keep the changes small and leave this be.

File: src/select-list-view.ts
Comment (winstliu) [inline src/select-list-view.ts]:
```suggestion
```

**Review by winstliu — APPROVED** [review]
I'm happy with this 👍.

**Comment by aminya** [comment]
@50Wliu Could you merge this if you have access? I don't want to wait another year for merging 😞

**Comment by winstliu** [comment]
Sorry, I don't merge things anymore :/.
@sadick254 is this something in your area?

**Pull request merged**
